Basic translation functionality

andyinabox opened the pull request:
This incorporates some very basic, targeted translation which can be edited in this Google Sheet: https://docs.google.com/spreadsheets/d/1m5QPNP6O8nsQsqJcQbwib_obpXRjmbYYGtwKzzg1l38/edit?pli=1#gid=0 (slack me and I'll give you edit access)

It's pretty rough, just trying to get something in there for now. Feel free to add comments/changes since this is really just a rough draft.

Resolves #40

andyinabox commented:
@hinzed1127 I didn't include your suggestion for auto language detection yet, figured that could be a next step

jdalt commented:
@andyinabox Can you elaborate on the strategy you're taking here and what changes will be necessary in the spreadsheet to accommodate them? How might this affect QA in the future?

jdalt commented on index.html:
```diff
@@ -32,7 +32,8 @@
 		mapboxgl.accessToken = 'pk.eyJ1IjoiamFjb2JyZGFsdG9uIiwiYSI6ImNrYXc2anFjbjIxNGwyeG14dWk0MDVycmIifQ.yILWplracVMB4mv1dWwTtg'
 		const DATA_URL = 'https://spreadsheets.google.com/feeds/list/1CyPozeKhmOuIaVnKDQAUKKsvA9R4F20hEQ3MSGWczP8/1/public/full?alt=json'
-	</script>
+		const TRANS_URL = 'https://spreadsheets.google.com/feeds/list/1m5QPNP6O8nsQsqJcQbwib_obpXRjmbYYGtwKzzg1l38/1/public/full?alt=json'
```
Is this basically a copy of the original spreadsheet that is translated whole hog? Are just parts translated?

experimatt replied:
This is a nitpick but I think using `TRANSLATION_URL` for the variable name, while longer, would be more appropriate.

andyinabox replied:
@jdalt only static labels like column headers, map legend, etc are being translated. Translating the actual data is a huge lift but this gets us a step closer

jdalt commented on index.html:
```diff
@@ -41,10 +42,17 @@
+        <li><a data-lang="spa" href="/?lang=spa">Español</a></li>
+        <li><a data-lang="kar" href="/?lang=kar">Karen</a></li>
+        <li><a data-lang="som" href="/?lang=som">Soomaali</a></li>
+        <li><a data-lang="hmn" href="/?lang=hmn">Hmong</a></li>
```
It sounds like we might have folks that can do Amharic and Oromo too. Doesn't have to be added here. Can be a followup - just want you to know.

andyinabox replied:
@experimatt I am using https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes verses the more common ones like `en` or `en-US` to support languages like Hmong and Karen. I wasn't able to find standard 2-letter language codes for those.

experimatt replied:
@andyinabox ahhhh gotcha, I didn't realize that. carry on then :)

jdalt replied:
til another iso code

kmarekspartz commented:
Clicking Español didn't change the text for me, on Mac Safari.

jdalt commented on translator.js:
```diff
@@ -0,0 +1,92 @@
+    this.els = document.querySelectorAll(`[${this.attr}]`) 
+    
+    let current
+    for (let i = 0; i < this.els.length; i++) {
```
Is there any reason this can't be done in the functional `els.forEach(` style?

andyinabox replied:
`querySelectorAll` returns a `NodeList` (not an array) which doesn't implement `forEach.` I'll convert to an array instead and use `forEach` for code consistency

jdalt replied:
silly dom api

kmarekspartz commented:
Syntax errors, looks similar to #28 

[Error] SyntaxError: Unexpected token '='. Expected an opening '(' before a method's parameter list.
	(anonymous function) (translator.js:16)
[Error] Error translating page – ReferenceError: Can't find variable: Translator
ReferenceError: Can't find variable: Translator
	(anonymous function) (script.js:279)
	asyncFunctionResume
	(anonymous function)
	promiseReactionJobWithoutPromise

jdalt commented on translator.js:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  // this converts a `gsx$keyname` to `keyname
+  convertGoogleKey(key) {
```
Is this just stubbed out or is it overridden somewhere? It's not clear to me why this is returning null or undefined.

andyinabox commented:
@jdalt no changes are needed to the spreadsheet, I am only translating labels here ... not data from the spreadsheet (that is a much bigger job). I left some notes about implementation here: https://github.com/Twin-Cities-Mutual-Aid/twin-cities-aid-distribution-locations/issues/40

@kmarekspartz thanks, hadn't tested in Safari yet!

cwood821 commented:
This is awesome!

Someone could disagree with me here, one tweak might be to make more of the Spanish filters noun phrases. For example, “Necesita dinero” > “necesidad de dinero”, “necessita voluntarias” > “necesidad de voluntarios”. That would line up with some of the existing translations like “estado de ubicacion”.

andyinabox commented:
@cwood821 I think we will probably have another process for vetting the actual translations ... I just used Google translate so I could show an example

andyinabox commented:
Push fixes/adjustments based on comments above. Some highlights:

 * Hoping I fixed the Safari issue @kmarekspartz mentioned, although cant test currently. Can anyone verify?
 * Switched to use a `<select>` for languages as @experimatt suggestion. I think it works ok, but my one hesitation is that it might be less visible. Before you could see "Soomaali" right on the page, but now you have to know to click the language dropdown first

A few additional notes

 * Didn't add any flag icons, partially because some of the languages (e.g. Hmong) don't have a corresponding flag
 * The legend/title area is getting pretty bloated but I think that would best be handled in a separate UI task
 * Currently disabled all languages except English and Spanish since we don't have translations yet for the rest (maybe Spanish shouldn't even be enabled since it's just from Google translate)

jdalt commented:
Safari looks good.

kmarekspartz commented:
Confirmed, works on Safari.

I'm not sold on the dropdown... If I'm not an English speaker, how would I know to click on the Language: English button to find e.g. Hmoob?

jdalt commented:
I really want to merge this but the mobile experience is hurting right now. Here's the chrome simulated iphone 5/6se view:
<img width="262" alt="Twin_Cities_Aid_Distribution_Locations" src="https://user-images.githubusercontent.com/2358067/83726547-49b90480-a609-11ea-98c7-ba8a20ce794d.png">

It's sort of unfair because you didn't create this problem but it's definitely worse with another vertical entry.

Let's talk in the morning about ways that we can improve the mobile experience.

kmarekspartz commented:
Other thought: feature toggle url parameter for testing while the translations get built out?

–
Kyle Marek-Spartz

> On Jun 4, 2020, at 02:18, Jacob Dalton <notifications@github.com> wrote:
> 
> ﻿
> I really want to merge this but the mobile experience is hurting right now. Here's the chrome simulated iphone 5/6se view:
> 
> 
> It's sort of unfair because you didn't create this problem but it's definitely worse with another vertical entry.
> 
> Let's talk in the morning about ways that we can improve the mobile experience.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

andyinabox commented:
Saman had a good suggestion in #57 

> Consider handling translation with an immediate popup the first time user visits the site that asks user for preferred language, and then saving the result to localstorage. We can then put the language toggle behind help/info.

andyinabox commented:
Update on this: it's working, although I haven't tested extensively or really been able to do a once-over to clean up the code. 

Main new feature here is a "welcome" modal that shows up if the user hasn't visited before.

I have to go back to work, but wanted to get this up there for review. Things that I'm aware of that still need to be done:

1. Refine styling for welcome modal (this could be done by somebody else)
2. Fix merge conflicts (probably needs to be me)
3. Move translation drop-down to "info" section (probably me, after fixing merge conflicts)
4. QA (can be done by somebody else)
5. Add translations (can be somebody else)

experimatt commented:
I can take a stab at resolving the merge conflicts, seeing as I just added some code that re-organized a few things

experimatt commented:
Merge conflicts resolved. We're going to need to add some additional rows to the translation spreadsheet to reflect additional content on the site that's been added recently.

samanpwbb commented on welcome.js:
```diff
@@ -0,0 +1,45 @@
+
+    return `
+      <div class="modal">
+        <h1 class="welcome-message">Welcome!</h1>
```
A snappy one-liner/purpose statement would be useful here I think:

```suggestion
        <h1 class="welcome-message">Welcome! Twin Cities Mutual Aid is an up to date resource on mutual aid sites in the Twin Cities.</h1>
```

experimatt replied:
@samanpwbb good call, I just added this!

samanpwbb commented on welcome.js:
```diff
@@ -0,0 +1,45 @@
+    const buttons = languages.map(l => `<button class="language-button" value="${l.name}">${l.label}</button>`).join('')
+
+    return `
+      <div class="modal">
```
Would it be possible to make it clear that this is a modal / hint at the content behind the modal by adding margin around the edges & not making the modal 100% height?

experimatt replied:
Good idea

andyinabox replied:
Yeah good idea, thanks!

benjifs commented:
Related to this PR but could be left for later:

We could also replace the labels in the popups from the map by adding `data-translation-id='${key}'` and adding those translations from the map.

https://github.com/Twin-Cities-Mutual-Aid/twin-cities-aid-distribution-locations/blob/a41acaf23bed16fde0745ce8c16350167dc58201/script.js#L260-L264

Since its currently using the `camelToTitle` function, it would just be adding those keys to the table and translating them. We would need to trigger `translate` on popup appearing as well which is why maybe we can just leave this for another issue.

Unrelated but the keys there have some spelling issues so a quick typo fix there would be great.

https://github.com/Twin-Cities-Mutual-Aid/twin-cities-aid-distribution-locations/blob/a41acaf23bed16fde0745ce8c16350167dc58201/script.js#L233-L239

andyinabox commented:
Ok, one more round on this one.

 * Preview with only translated languages visible: https://deploy-preview-54--wonderful-montalcini-96d80d.netlify.app/
 * Preview with all languages visible: https://deploy-preview-54--wonderful-montalcini-96d80d.netlify.app/?dev

The UX has change significantly, how y'all are onboard for the updates.

1. Still doing the "Welcome" modal when user first visits
2. Language is displayed in the top left corner of the map
3. Click on the language will re-open the "Welcome" modal

thanks to @experimatt for diving in on this when I wasn't available!

andyinabox commented:
@benjifs yeah, I've had that in the back of my mind but was trying to keep things more limited to get out the door. But probably makes sense to add all the headers from the main data source to the translation doc, so we don't have to go back to the translators more than once

benjifs commented:
@andyinabox just added the keys and English and Spanish translations for those. I'm also adding some changes in a different branch which would add the translations to the map popups

samanpwbb commented:
I can do a really quick design cleanup/consistification pass here in about 20 mins.

andyinabox commented:
@benjifs Awesome! Thanks for doing that. Couldn't sleep, so I'm gonna make the change you suggested :grin:. Will probably switch the ids to snake_case for consistency

samanpwbb commented:
I hit this error every time when running from localhost and at the staging URL – is this a known issue?

![Screen Shot 2020-06-04 at 9 55 55 PM](https://user-images.githubusercontent.com/108094/83838496-3318b900-a6ae-11ea-9f26-816e17632728.png)

samanpwbb commented:
Just pushed a really quick design cleanup (added max width to modal, cleaned up spacing):

![Screen Shot 2020-06-04 at 9 58 21 PM](https://user-images.githubusercontent.com/108094/83838608-93a7f600-a6ae-11ea-85d2-0e45fa99653c.png)

And on desktop:
![Screen Shot 2020-06-04 at 9 58 13 PM](https://user-images.githubusercontent.com/108094/83838621-9e628b00-a6ae-11ea-96b3-706c786f5052.png)

benjifs commented on translator.js:
```diff
@@ -40,35 +49,44 @@ class Translator {
+   */
   get language() {
     const lang = window.localStorage.getItem(Translator.LOCAL_STORAGE_KEY)
-    if (this.validateLanguageKey(lang)) return lang
```
@samanpwbb This line is causing the issue. If I add it back it works for me

andyinabox commented:
@samanpwbb oh yeah sorry, I did a cleanup pass and was a little overzealous! Added a comment explaining why that line is necessary.

Just pushed @benjifs's suggestion for setting the map popup labels. Also fixed issue where sidebar cards weren't being translated on filter change.

andyinabox commented:
also those design tweaks look great!

experimatt commented:
This looks so great, amazing work everyone!

experimatt commented on script.js:
```diff
@@ -160,7 +210,9 @@ const createListItem = (location, status, lng, lat) => {
         essential: true,
         zoom: 13
       })
+      console.log('popup', popup)
```
errant console I wonder if we can remove this

experimatt replied:
Couldn't get this console.log to run locally, so I left it in. :yolo:

experimatt commented:
I just tested this locally and in netlify and think it looks great. merging!

experimatt commented:
We reverted this, then unreverted it once we cleaned up the underlying data issues. whoops
